src/compiler/checker.ts

```diff
@@ -4825,6 +4763,88 @@ namespace ts {
             return widened;
         }
 
+        function getJSDocTypeFromSpecialDeclarations(declaredType: Type | undefined, expression: Expression, symbol: Symbol, declaration: Declaration) {
+            const typeNode = getJSDocType(expression.parent);
+            if (typeNode) {
+                const type = getWidenedType(getTypeFromTypeNode(typeNode));
+                if (!declaredType) {
+                    return type;
+                }
+                else if (declaredType !== errorType && type !== errorType &&
+                         !isTypeIdenticalTo(declaredType, type) &&
+                         !(symbol.flags & SymbolFlags.JSContainer)) {
+                    errorNextVariableOrPropertyDeclarationMustHaveSameType(declaredType, declaration, type);
+                }
+            }
+            return declaredType;
+        }
+
+        /** If we don't have an explicit JSDoc type, get the type from the initializer. */
+        function getInitializerTypeFromSpecialDeclarations(symbol: Symbol, resolvedSymbol: Symbol | undefined, expression: Expression, special: SpecialPropertyAssignmentKind) {
+            if (isBinaryExpression(expression)) {

```

Comment:
Might be neater to do this check outside and type the parameter as `expression: BinaryExpression`.

Reply:
Putting the check as an inline `isBinaryExpression(expression) ? getInitializer...` is a bit unwieldy, and I expect that we may end up doing some real work in the non-binary expression-case instead of just `return neverType`. (You need to return some type so that `types.length === symbol.declarations.length` in getConstructorDefinedThisAssignmentTypes.)